### PR TITLE
Some minor changes to ease implementation on top of kontiki

### PR DESCRIPTION
**Pull request by abailly**
This PR contains a bunch of changes I had to make to kontiki in order to implement [raptr](https://github.com/capital-match/raptr):
- reset timeouts even when they are ignored by current state
- store last known leader within follower's state

I also moved tests to more standard test/ directory and reorganized them to use tasty. 

I can split this PR in smaller chunks if required.

### CODE REVIEW

**Comment by NicolasT** [comment]
I'm sorry I completely missed this PR (and it's been  a year... :-/)
I'll go over it right away.

File: bin/demo.hs [hunk 1]
```diff
@@ -1,37 +1,14 @@
...
-    logLastEntry = do
-        l <- ask
-        return $ if IntMap.null l
-                    then Nothing
-                    else Just $ snd $ IntMap.findMax l
-
-runMemLog :: MemLog a r -> Log a -> r
-runMemLog = runReader . unMemLog
+import qualified Data.IntMap          as IntMap
+import           Data.Kontiki.MemLog
```
Comment (NicolasT) [inline bin/demo.hs]: See later comment

File: kontiki.cabal [hunk 2]
```diff
@@ -21,32 +21,33 @@ Source-Repository head
 
 Library
   Exposed-Modules:     Network.Kontiki.Raft
+                     , Data.Kontiki.MemLog
```
Comment (NicolasT) [inline kontiki.cabal]: I'm not very fond of this. The `MemLog` implementation is something that can be useful in tests, and may be useful in some demo tools, but it defies all guarantees a system like Raft needs from its log implementation w.r.t. (crash) safety.

As such, I don't think this module should be part of the library and definitely not be exposed as part of its API to any users.

File: kontiki.cabal [hunk 3]
```diff
@@ -21,32 +21,33 @@ Source-Repository head
...
                      , Network.Kontiki.Raft.Leader
                      , Network.Kontiki.Raft.Utils
-  Build-Depends:       base >= 4.8 && < 4.9
-                     , mtl >= 2.2 && < 2.3
-                     , bytestring >= 0.10 && < 0.11
-                     , containers >= 0.5 && < 0.6
-                     , binary >= 0.7 && < 0.8
-                     , lens >= 4.12 && < 4.13
-                     , QuickCheck >= 2.8 && < 2.9
+  Build-Depends:       base
```
Comment (NicolasT) [inline kontiki.cabal]: I really like to have version bounds in dependency definitions. The bounds may be too restrictive (the code is getting quite old...), but just dropping them makes publishing this package to Hackage or usage by others more difficult.

File: kontiki.cabal [hunk 4]
```diff
@@ -55,33 +56,33 @@ Executable kontiki-udp
...
 Test-Suite kontiki-test
   Type:                exitcode-stdio-1.0
-  Main-Is:             test.hs
+  Main-Is:             kontiki-test.hs
   Build-Depends:       base
                      , mtl
                      , binary
-                     , test-framework >= 0.8 && < 0.9
-                     , test-framework-quickcheck2 >= 0.3 && < 0.4
+                     , test-framework
```
Comment (NicolasT) [inline kontiki.cabal]: Is `test-framework` etc still used given your changes?

File: src/Data/Kontiki/MemLog.hs [hunk 5]
```diff
@@ -10,12 +10,12 @@ module Data.Kontiki.MemLog (
     , IntMap.insert
     ) where
 
-import Data.IntMap (IntMap)
-import qualified Data.IntMap as IntMap
+import           Data.IntMap          (IntMap)
```
Comment (NicolasT) [inline src/Data/Kontiki/MemLog.hs]: Des gouts et des couleurs... but I don't like this formatting ;)

File: src/Network/Kontiki/Raft.hs [hunk 6]
```diff
@@ -38,7 +38,7 @@ import qualified Network.Kontiki.Raft.Leader as Leader
 -- and `state' of the node, runs the Raft protocol and  
 -- returns the new state of the node and a list of commands that 
 -- should be executed by the driver.
-handle :: (Functor m, Monad m, MonadLog m a) 
+handle :: (Monad m, MonadLog m a)
```
Comment (NicolasT) [inline src/Network/Kontiki/Raft.hs]: This removes compatibility with older compilers. The 'Redundant Constraints' warning of GHC8.0.1 was basically a mistake and is (for good reason) being removed from `-Werror` again, if I'm not mistaken (check GHC Trac). Would it be possible to add `-fno-warn-redundant-constraints` to the GHC options in the Cabal file instead (under a version check)?

File: src/Network/Kontiki/Raft/Candidate.hs [hunk 7]
```diff
@@ -123,14 +123,15 @@ handleElectionTimeout = do
                                  }
 
 -- | Handles `HeartbeatTimeout'.
-handleHeartbeatTimeout :: (Functor m, Monad m)
+handleHeartbeatTimeout :: (Monad m)
                        => TimeoutHandler HeartbeatTimeout a Candidate m
 handleHeartbeatTimeout = do
+    resetHeartbeatTimeout
```
Comment (NicolasT) [inline src/Network/Kontiki/Raft/Candidate.hs]: Are you sure this is the correct thing to do? We don't have to send a heartbeat as long as we're Candidate, so why resetting its timeout?

File: src/Network/Kontiki/Raft/Follower.hs [hunk 8]
```diff
@@ -194,14 +196,15 @@ handleElectionTimeout = do
         else Candidate.stepUp nextTerm commitIndex
 
 -- | Handles `HeartbeatTimeout'.
-handleHeartbeatTimeout :: (Functor m, Monad m)
+handleHeartbeatTimeout :: (Monad m)
                        => TimeoutHandler HeartbeatTimeout a Follower m
 handleHeartbeatTimeout = do
+    resetHeartbeatTimeout
```
Comment (NicolasT) [inline src/Network/Kontiki/Raft/Follower.hs]: Same question as for the change in Candidate.

File: src/Network/Kontiki/Raft/Leader.hs [hunk 9]
```diff
@@ -107,12 +107,15 @@ quorumIndex = do
     return $ sorted !! (quorum - 1)
 
 -- | Handles `ElectionTimeout'.
-handleElectionTimeout :: (Functor m, Monad m)
+handleElectionTimeout :: (Monad m)
                       => TimeoutHandler ElectionTimeout a Leader m
-handleElectionTimeout = currentState
+handleElectionTimeout =  do
+  resetElectionTimeout
```
Comment (NicolasT) [inline src/Network/Kontiki/Raft/Leader.hs]: Similar question as the similar heartbeat timeout change in Candidate or Follower state

File: src/Network/Kontiki/Raft/Utils.hs [hunk 10]
```diff
@@ -1,7 +1,4 @@
-{-# LANGUAGE TypeFamilies,
-             ScopedTypeVariables,
-             FlexibleContexts,
-             OverloadedStrings #-}
+{-# LANGUAGE FlexibleContexts, OverloadedStrings, ScopedTypeVariables, TypeFamilies #-}
```
Comment (NicolasT) [inline src/Network/Kontiki/Raft/Utils.hs]: This is a different formatting than the other changes.

File: src/Network/Kontiki/Raft/Utils.hs [hunk 11]
```diff
@@ -55,17 +52,17 @@ handleGeneric
     handleHeartbeatTimeout
     event = case event of
     EMessage s m -> case m of
-        MRequestVote m' -> handleRequestVote s m'
-        MRequestVoteResponse m' -> handleRequestVoteResponse s m'
-        MAppendEntries m' -> handleAppendEntries s m'
+        MRequestVote m'           -> handleRequestVote s m'
```
Comment (NicolasT) [inline src/Network/Kontiki/Raft/Utils.hs]: Hmpf.

File: test/Network/Kontiki/RaftSpec.hs [hunk 12]
```diff
@@ -0,0 +1,12 @@
+module Network.Kontiki.RaftSpec where
```
Comment (NicolasT) [inline test/Network/Kontiki/RaftSpec.hs]: I'm very confused by this module in general. What is it testing? Why is Test.QuickCheck, MemLog,... imported?

File: test/kontiki-test.hs [hunk 13]
```diff
@@ -0,0 +1,16 @@
+{-# LANGUAGE FlexibleInstances          #-}
```
Comment (NicolasT) [inline test/kontiki-test.hs]: Since all tests use HSpec, any reason to use Tasty at all?

**Review by NicolasT — CHANGES_REQUESTED** [review]
Several questions, and some things I think definitely need to change:
- Don't only support AMP compilers, ideally
- Don't expose MemLog
- Keep dependency version bounds

**Comment by abailly** [comment]
Thanks for the comments, I will address them.

**Pull request closed without merging**